libfoundation/src/foundation-java.cpp

```diff
@@ -0,0 +1,199 @@
+
+MC_DLLEXPORT_DEF MCTypeInfoRef MCJavaGetObjectTypeInfo() { return kMCJavaObjectTypeInfo; }
+
+struct __MCJavaObjectImpl
+{
+    void *object;
+};
+
+__MCJavaObjectImpl *MCJavaObjectGet(MCJavaObjectRef p_obj)
+{
+    return static_cast<__MCJavaObjectImpl *>(MCValueGetExtraBytesPtr(p_obj));
+}
+
+static inline __MCJavaObjectImpl MCJavaObjectImplMake(void* p_obj)
+{
+    __MCJavaObjectImpl t_obj;
+    t_obj.object = p_obj;
+    return t_obj;
+}
+
+static void __MCJavaObjectDestroy(MCValueRef p_value)
+{
+    MCJavaObjectRef t_obj = static_cast<MCJavaObjectRef>(p_value);
+    MCJavaPrivateDestroyObject(t_obj);
+}
+
+static bool __MCJavaObjectCopy(MCValueRef p_value, bool p_release, MCValueRef &r_copy)
+{
+    if (p_release)
+        r_copy = p_value;
+    else
+        r_copy = MCValueRetain(p_value);
+    return true;
+}
+
+static bool __MCJavaObjectEqual(MCValueRef p_left, MCValueRef p_right)
+{
+    if (p_left == p_right)
+        return true;
+    
+    return MCMemoryCompare(MCValueGetExtraBytesPtr(p_left), MCValueGetExtraBytesPtr(p_right), sizeof(__MCJavaObjectImpl)) == 0;
+}
+
+static hash_t __MCJavaObjectHash(MCValueRef p_value)
+{
+    return MCHashBytes(MCValueGetExtraBytesPtr(p_value), sizeof(__MCJavaObjectImpl));
+}
+
+bool __MCJavaObjectDescribe(MCValueRef p_value, MCStringRef &r_desc)
+{
+    if (!s_java_initialised)
+        return false;
+
+    return MCJavaPrivateObjectDescribe(p_value, r_desc);
+}
+
+static MCValueCustomCallbacks kMCJavaObjectCustomValueCallbacks =
+{
+    true,
+    __MCJavaObjectDestroy,
+    __MCJavaObjectCopy,
+    __MCJavaObjectEqual,
+    __MCJavaObjectHash,
+    __MCJavaObjectDescribe,
+    
+    nullptr,
+    nullptr,
+};
+
+MC_DLLEXPORT_DEF bool MCJavaObjectCreate(void *p_object, MCJavaObjectRef &r_object)
+{
+
+    MCJavaObjectRef t_obj;
+    if (!MCValueCreateCustom(kMCJavaObjectTypeInfo,
+                             sizeof(__MCJavaObjectImpl), t_obj))
+        return false;
+
+    *MCJavaObjectGet(t_obj) = MCJavaObjectImplMake(p_object);
+    r_object = t_obj;
+
+    return true;
+}
+
+bool MCJavaCreateJavaObjectTypeInfo()
+{
+    return MCNamedCustomTypeInfoCreate(MCNAME("com.livecode.java.JavaObject"), kMCNullTypeInfo, &kMCJavaObjectCustomValueCallbacks, kMCJavaObjectTypeInfo);
+}
+
+MC_DLLEXPORT_DEF void *MCJavaObjectGetObject(const MCJavaObjectRef p_obj)
+{
+    __MCJavaObjectImpl *t_impl = MCJavaObjectGet(p_obj);
+    return t_impl -> object;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+MC_DLLEXPORT_DEF
+bool MCJavaCheckSignature(MCTypeInfoRef p_signature, MCStringRef p_args, MCStringRef p_return, int p_call_type)
+{
+    return MCJavaPrivateCheckSignature(p_signature, p_args, p_return, p_call_type);
+}
+
+MC_DLLEXPORT_DEF
+bool MCJavaCallJNIMethod(MCNameRef p_class_name, void *p_method_id, int p_call_type, MCTypeInfoRef p_signature, void *r_return, void **p_args, uindex_t p_arg_count)
+{
+    if (!s_java_initialised)
+        return false;
```

Comment:
It might be a good idea to `MCErrorThrow()` something if code tries to do Java stuff without initialising Java. At the moment it will just fail silently.